avoid unnecessary dependencies between save tasks in dataframe.to_hdf

nirizr opened the pull request.

mrocklin commented on dask/dataframe/io.py:
```diff
@@ -430,30 +430,60 @@ def to_hdf(df, path_or_buf, key, mode='a', append=False, complevel=0,
-                             'fletcher32': fletcher32})))
+    task = (apply, pd_to_hdf,
+               (tuple, [(df._name, 0), fmt_obj(path_or_buf, i_name),
+                   key.replace('*', i_name)]),
```
We can probably drop the delayed call to `tuple`.  The schedulers have evolved a bit since this was first written

mrocklin commented on dask/dataframe/io.py:
```diff
@@ -430,30 +430,60 @@ def to_hdf(df, path_or_buf, key, mode='a', append=False, complevel=0,
+                   'fletcher32': fletcher32}))
+        if link:
+            linked_key = (name, 0) if link == 'first' else (name, i - 1)
+            task = (_link, linked_key, task)
```
Nice.  That's a pretty clean way of supporting both cases.

mrocklin commented:
If you wanted to test behavior you could either explicitly check for the existence of `link` in the graph *or* you could compute the dependencies of all keys in the graph and make sure that the max was 1.  The function `dask.core.get_deps` might be of interest here.

nirizr commented:
Thanks for the comments, i'll try adding some relevant tests during the weekend.

mrocklin commented on dask/dataframe/io.py:
```diff
@@ -406,65 +406,81 @@ def _link(token, result):
+                     key.replace('*', i_name)], kwargs2))
+        if single_file or single_node:
+            dep_link = (name, i - 1 if single_node else 0)
+            task = (_link, dep_link, task)
```
If looks like we link twice here

mrocklin replied:
Once on this line and once on the original definition of `task` up above

nirizr commented:
Fixed the double-linking mistake, added more tests.
Still not satisfied with the tests.

nirizr commented:
There is one issue i'm bothered with in this solution. it shouldn't prevent merging this solution when i'm done but i feel it should be noted someplace.

when saving to multiple nodes in the same hdf file (which i feel is the best compromise between efficiency and clean output), there's a bottleneck created by all subsequent to_hdf tasks depending on the first.
It might be worth creating a fake "to_hdf" node that will not depend on any input for this case.

mrocklin commented:
That seems reasonable to me

nirizr commented:
Add `max_depth` function to `dask.core`. it uses `get_deps` to calculate the height of the dependency tree. Its isn't too efficient (won't cache already calculated subtree heights) but it works.
It might warrant it's own tests, now that i think about it.

I'm using the max_depth with to_hdf(compute=False) to calculate the height of the dependency tree to test the number of links in the graph.

mrocklin commented:
I recommend that we move `max_depth` to `utils.py`, perhaps just call it `depth`, and put a `toolz.memoize` decorator on the inner function to help with caching.  In order to maintain stdlib only on the core lets import `toolz` within the `max_depth` function rather than at the utils.py module level.

nirizr commented:
depth feels too general, what do you think about dask_depth or graph_depth? dep_depth?

I thought because core has all the other dask graph related functions it's a better place, i'll move it once we agree on a name.

nirizr commented:
Moving `depth` to `utils.py` will require importing `get_deps` from `core.py`, making sure that's ok because currently `utils.py` imports only basics and from `compatibility.py`.

mrocklin commented:
I'm fine with `max_depth` or whatever you prefer.

I like to keep `core.py` minimal where possible.  At the moment `max_depth` is only used for testing and not actual operation.

nirizr commented:
renamed to `dependency_depth` and moved to `utils.py`.
Hope it's alright.

Any more comments? additional tests maybe?

mrocklin commented:
This all seems fine to me

nirizr commented:
I fear saving to a single file in parallel might result in corruptions, as it's not supported by hdf.
Any ideas how we can easily test this?

passing lock=True to pandas might help, but we don't force or check for that right now.

I'll try testing this at work with a medium unbalanced dataset, hoping it'll reveal those issues.

nirizr commented:
I was positive pandas or dask had a lock parameter and will handle it if passed, but i can't seem to find it now.
Looks like i'll need to implement a lock before this gets too useful.
it seems that both multiprocessing and multithreading async saves break without a lock.

nirizr commented:
I've had some trouble fixing segfaults with the current to_hdf tests layout, so i moved some tests a bit and added more tests.
I also made a better default get selector, and commented the reasoning behind it.

writing to multiple hdf files is now optimized (using `multiprocessing.get`).
writing to multiple nodes simplifies the `_link` chain so we're able to write partitions as they're ready (instead waiting for all previous partitions), but scheduler is still `sync` until we have locking available.
writing to a single node is just as it was (defaults to `sync`, lots of `_link`s).

we can either merge this as is (after adding a NotImplemented) on any value for `lock` or implement the locking in this PR.
Either way works for me, with a slight tendency towards the former as it carries enough improvements, birds at hand vs bush etc.

mrocklin commented on dask/dataframe/io.py:
```diff
@@ -406,65 +407,96 @@ def _link(token, result):
+        if not single_file:
+            get = dask.multiprocessing.get
+        else:
+            get = get_sync
```
I propose that the two options here should be "synchronous if everything is linked, or whatever the default was before otherwise".  This is generally done by `get=None`

nirizr replied:
That was what I was going for initially, but the default (dask.threaded.get) ends up segfaulting when trying to write to multiple files.

mrocklin replied:
It seems like the right solution there is to handle locks.

These days I use the distributed scheduler on a single machine as often as threaded or multiprocesing, so I'm somewhat hesitant to hard code in threaded or multiprocessing.

mrocklin commented on dask/dataframe/tests/test_io.py:
```diff
@@ -758,6 +783,169 @@ def test_to_hdf_multiple_datasets():
+        a = a.apply(delayed_nop, axis=1, columns=a.columns)
+        a.to_hdf(fn, '/data*')
+        out = dd.read_hdf(fn, '/data*')
+        eq(df16, out)
```
I don't understand this test.  Why are we adding delays?

nirizr replied:
It was a simple way to make sure results don't arrive in their expected order. I can speed it up by only delaying some of them.

nirizr replied:
Do you think this test is redundant?

mrocklin replied:
It makes sense to me

mrocklin commented:
Probably `to_hdf` never got the lock treatment because it was synchronous until now.  I would look towards `dask.array.core.store` for current conventions on concurrent writes with locks.

nirizr commented:
Would you prefer delaying this PR until I add locks or should I remove the hardcoded multiprocessing and implement the locks in another PR after this one is merged?

mrocklin commented:
Personally I see significantly more value if this has the locks included.  I can also add these if preferred.

nirizr commented:
Removed hardcoded `dask.multiprocessing.get` in favor of locks.
added tests for three schedulers available within dask (`get_sync`, `dask.threaded.get` and `dask.multiprocessing.get`).

There are two issues i'd like guidance for:

1. I couldn't use locks with the multiprocessing scheduler, and that seems to be the case with `dask.array.core.store` as well, so I explicitly used `lock=False` in those. We might want to save the trouble for the user and make `lock=False` the default when the multiprocessing scheduler is used.

2. Obviously `multiprocessing.get` won't work without a lock to when writing to multiple nodes in a single file (simultaneous writes) so i marked that test as expected fail. I also added a similar expected fail test to `dask.array.core.store`.
These tests sometimes fail with hdf exceptions and sometimes fail silently with erroneous results, so i'll appreciate any fix suggestions and if nothing comes we'll need to explicitly check of this case and raise an exception in `to_hdf`.

mrocklin commented on dask/dataframe/io.py:
```diff
@@ -447,27 +447,42 @@ def to_hdf(df, path_or_buf, key, mode='a', append=False, complevel=0,
+                lock.release()
+                print("released")
+
+        return None
```
If possible this function should be moved outside of the function so that it is easier to serialize.  We should probably also remove the go and print logic.

nirizr replied:
Sorry, the prints an go were obviously leftovers (from trying to use multiprocessing.Lock, actually).

mrocklin replied:
Dask.dataframe depends on partd which depends on locket.  You might want to
try locket.lock_file

On Mon, Jun 27, 2016 at 2:39 PM, Nir <notifications@github.com> wrote:

> In dask/dataframe/io.py
> <https://github.com/dask/dask/pull/1293#discussion_r68661234>:
>
> > +        if lock:
> > +            go = False
> > +            while not go:
> > +                print(lock)
> > +                go = lock.acquire(blocking=False)
> > +                print(go)
> > +            #lock.acquire()
> > +            print("acquired")
> > +        try:
> > +            pd_to_hdf(*args, **kwargs)
> > +        finally:
> > +            if lock:
> > +                lock.release()
> > +                print("released")
> > +
> > +        return None
>
> Sorry, the prints an go were obviously leftovers (from trying to use
> multiprocessing.Lock, actually).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/1293/files/4e7f0722af962655360b5c7b8b6b1b869737b5fa..62c7ecd0f731fd39bec25e563e244daf69158bfd#r68661234>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AASszLF-uND7l7CCQ_smChNi-fnTBm7Pks5qQEL1gaJpZM4I2sWs>
> .
>

mrocklin commented on dask/dataframe/tests/test_io.py:
```diff
@@ -959,6 +959,97 @@ def test_to_hdf_exceptions():
+    with tmpfile('h5') as fn:
+        a.to_hdf(fn, '/data*', get=dask.multiprocessing.get)
+        out = dd.read_hdf(fn, '/data*')
+        eq(df, out)
```
You could try a `multiprocessing.Lock` instead or a file based lock such as is found in the `locket` library.

nirizr replied:
Tried, didn't work so good, will try some more.

mrocklin commented:
You're correct that `threading.Lock` objects will fail in multiprocessing.  There are multiprocessing-friendly locks around though.  This was the motivation behind passing an explicit lock into the `da.store` function.  It allows informed users to use a lock of their choice.

nirizr commented:
`locket.lock_file` uses `threading.Lock` so it's not an easy option.
I'll see what else i can do.

mrocklin commented:
We could also just stick with `lock=False` and leave this concern for the user, perhaps warning if they're using multiprocessing in an unsafe way?

nirizr commented:
Yes, when i'll give up i'll add a warning/exception. any preference btw?
Also, should i only check against manually specified multiprocessing.get or should it warn when the default was changed using `set_options`?

I also wonder if this effects dask.distributed.

nirizr commented:
there's a third possibility: changing how `_link` is used in that case (multiprocess to single file with multiple nodes), although it might get messy

nirizr commented:
I think locks are good now. Failure is due to boto3 update.

We now find out if we're going to use the multiprocessing scheduler, and default to multiprocessing.Manager.Lock instead of Lock imported from threading.
Please make sure this is acceptable. If it is, it might be worth doing elsewhere?

mrocklin commented:
This looks good to me.  +1

nirizr commented:
Rebase + squash + add [forgotten](https://github.com/dask/dask/pull/1293/files#diff-2a4c3ed91854a1a70080d69581e762b3R16) include

ankravch commented:
I wonder if this issue has been resolved ?

nirizr commented:
This PR awaits it's merge, afaik. You (and anyone else, actually) are encouraged to suggest changes or offer comments.
I am currently testing this branch in some real use cases at work, and any additional experiments would be greatly appreciated.

This is a performance improvement and not a bugfix, so I won't consider merging it urgent.
I am however anxious to start using it as part of dask but It is my understanding that Matthew is a bit preoccupied this week and i see no reason to rush it.

mrocklin commented:
The delay here was mostly due to my absence last week.  Merging now.

nirizr commented:
Thanks!

ankravch commented:
I can confirm that to_hdf works as I would expect, thank you!
P.S. I wonder if there is/will be a dask.dataframe.to_bcolz(...) function ?

mrocklin commented:
@ankravch care to submit a PR for `to_bcolz`?
